### PR TITLE
Provide explicitly pausing annotations of ScaledObjects at the current replica count

### PR DESCRIPTION
**Pull request by SpiritZhou**
Add pausing annotations of ScaledObjects  at the current replica count

1. Refactor current paused checking method of ScaledObjects
2. Add autoscaling.keda.sh/paused:true in ScaledObjects

### Checklist

- [X] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [X] Tests have been added
- [X] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Relates to https://github.com/kedacore/keda/issues/3304
Relates to https://github.com/kedacore/keda-docs/pull/1190

### CODE REVIEW

File: tests/internals/pause_scaledobject_explicitly/pause_scaledobject_explicitly_test.go
Comment (tomkerkhove) [inline tests/internals/pause_scaledobject_explicitly/pause_scaledobject_explicitly_test.go:188]: Before we do this, can we check that it's still at 5 replicas?

Comment (tomkerkhove) [reply tests/internals/pause_scaledobject_explicitly/pause_scaledobject_explicitly_test.go:188]: Or you prefer doing after when it scales to 0 so that it should have scaled in if it was active, but is still at 5? If so, fine for me.

**Review by tomkerkhove — APPROVED** [review]
LGTM

**Comment by tomkerkhove** [comment]
Can you add an e2e test that applies both paused + paused with specific replica count and make sure that the explicit count wins please?

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
Can you add an e2e test that applies both paused + paused with specific replica count and make sure that the explicit count wins please?

**Comment by SpiritZhou** [comment]
Add  e2e test that applies both paused + paused.

File: tests/internals/pause_scaledobject_explicitly/pause_scaledobject_explicitly_test.go
Comment (SpiritZhou) [reply tests/internals/pause_scaledobject_explicitly/pause_scaledobject_explicitly_test.go:204]: Added checking to make sure it's still at 5 replicas

**Comment by SpiritZhou** [comment]
There is a Solr e2e test failed. It seems like that there is a problem when setting up Solr. Do you have any idea to fix it? @JorTurFer 

image

**Comment by JorTurFer** [comment]
> There is a Solr e2e test failed. It seems like that there is a problem when setting up Solr. Do you have any idea to fix it? @JorTurFer

Let me check. In any case, you can ignore it as it's not related with this PR

File: apis/keda/v1alpha1/scaledobject_types.go
Comment (zroubalik) [inline apis/keda/v1alpha1/scaledobject_types.go:179]: Could you please rename this method to be more clear what is it about? also could you please add a comment with short description of the functionality?

File: apis/keda/v1alpha1/scaledobject_types.go
Comment (zroubalik) [inline apis/keda/v1alpha1/scaledobject_types.go:173]: please add comment with description

**Comment by tomkerkhove** [comment]
@SpiritZhou while waiting for a review; would you mind updating our docs in a separate PR please?

**Comment by SpiritZhou** [comment]
> @SpiritZhou while waiting for a review; would you mind updating our docs in a separate PR please?

I have updated our docs (https://github.com/kedacore/keda-docs/pull/1190). Do you mean we need to add more detail info in the doc?

**Review by tomkerkhove — APPROVED** [review]
LGTM

**Comment by tomkerkhove** [comment]
Any concerns for you @zroubalik / @JorTurFer ?

**Review by zroubalik — APPROVED** [review]
LGTM, great job!

**Review by JorTurFer — COMMENTED** [review]
Looking good, just an small nit

**Review by zroubalik — APPROVED** [review]
LGTM, let's merge this once docs PR is finished.

Thanks for the contribution!

**Comment by SpiritZhou** [comment]
> LGTM, let's merge this once docs PR is finished.
> 
> Thanks for the contribution!

Hi @zroubalik do you mind to review the docs PR again?

**Review by zroubalik — APPROVED** [review]

**Pull request merged**
